### PR TITLE
fix(cypress): improve cypress performances

### PR DESCRIPTION
**Pull request by skjnldsv**
- [x] Prevent fetching all deps again with npm ci for each runner (node_modules is already saved in the init step)
- [x] Enable redis to speed things up
- [x] Allow restoring DB snapshots and implement it o the profile spec (`personal-info.cy.ts`)
- [x] Make Nextcloud setup into RAM

| Before | After |
|:---------:|:------:|
|||

### CODE REVIEW

**Comment by skjnldsv** [comment]
Also tested a new approach, restroring DB snapshots.
A good example is the profile test going from `3 minutes, 22 seconds` to ` 2 minutes, 36 seconds`

**Comment by SystemKeeper** [comment]
Not sure if this makes sense for cypress, but in talk we had some improvements when using apcu for local caching and setting the hashing method

https://github.com/nextcloud/spreed/blob/3d3ce29963ad99d8c17b76b947b436ee1effcd75/.github/workflows/integration-pgsql.yml#L149-L150

**Comment by susnux** [comment]
> we had some improvements when using apcu

Yes I think redis improves a bit here, but it adds network requests and does not make that much sense here as this is a local test with only one user at the same time.
So APCu as a local should bring another boost :)

**Comment by skjnldsv** [comment]
Thank you both, I'll adjust :)

**Comment by skjnldsv** [comment]
@susnux @SystemKeeper for apcu: https://github.com/nextcloud/docker-ci/pull/654

**Review by artonge — APPROVED** [review]

**Comment by skjnldsv** [comment]
Follow-up, use functions from `@nextcloud/cypress`

File: .github/workflows/cypress.yml
[the remark this answers is not in the document]
Comment (skjnldsv) [reply .github/workflows/cypress.yml:1]: https://github.com/nextcloud-libraries/nextcloud-cypress/pull/598

**Comment by susnux** [comment]
This is much better, thank you!
But there is still a performance issue, see runner 3:

Overall Run: 13 minutes 14 seconds
Cypress tests : 6 minutes 22 seconds
Pulling image: 1 minute
Restore context: 3 minutes

more than 50% of the CI time is still not related to tests.

But the thing that is weird: Creating the context takes ~2:30 while restoring takes ~3 minutes.
This are 27 minutes wasted by restoring the context, maybe this is something we can improve in a follow up?

File: cypress/e2e/settings/personal-info.cy.ts [hunk 1]
```diff
@@ -102,11 +102,23 @@ const genericProperties = ['Location', 'X (formerly Twitter)', 'Fediverse']
...
 		// ensure we can set locale and language
 		cy.runOccCommand('config:system:delete force_language')
 		cy.runOccCommand('config:system:delete force_locale')
+		cy.createRandomUser().then(($user) => {
+			user = $user
+			cy.modifyUser(user, 'language', 'en')
+			cy.modifyUser(user, 'locale', 'en_US')
+		})
+
+		cy.wait(500)
```
[the remark this answers is not in the document]
Comment (susnux) [reply cypress/e2e/settings/personal-info.cy.ts:117]: Why is this needed?
As soon as this part is reached the create user request has terminated so all database actions are already finished as the PHP process is done.
So normally this should not be needed I guess?

**Review by susnux — APPROVED** [review]
:rocket:

**Comment by skjnldsv** [comment]
> But the thing that is weird: Creating the context takes ~2:30 while restoring takes ~3 minutes.
> This are 27 minutes wasted by restoring the context, maybe this is something we can improve in a follow up?

It's decompressing the archive I think, it takes many IO.

> more than 50% of the CI time is still not related to tests.

That would be my guess as well, the more we have running, the more load our CI servers have. Meaning it will obviously have an impact on other runs.

**Comment by skjnldsv** [comment]
> Pulling the image takes locally for me much less time, I assume our runners do not have it cached and thus need to externally fetch it from GH?

It does, we have a docker layer cache. We store images.
But this one is a new image, and we have 4 CI servers. Maybe that was a cache miss occurence, should be better now :+1:

**Pull request merged**